lib/auth0_client.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+require 'jwt'
+require 'net/http'
+
+class Auth0Client
+  Error = Struct.new(:message, :status)
+  Response = Struct.new(:decoded_token, :error)
+
+  def self.domain_url
+    "https://#{Rails.application.config.x.auth0.domain}/"
```

Comment:
It might be safer for us to build up these URLs with the Ruby `URI::HTTPS` method: https://ruby-doc.org/stdlib-2.5.1/libdoc/uri/rdoc/URI/HTTP.html#method-c-build (these are the docs for `URI::HTTP`, but I think the API is the same for `HTTPS` but they just don't have good docs for them).

I'm mostly concerned about the trailing slash that you put here, since you're also making the assumption later in the file that the trailing slash exists so that you can join it with the `.well-known/` path. I've just hit lots of issues in the past with two different parts of a code base making incorrect assumptions about whether a trailing slash will exist or not, and these kinds of things all go away if you build up URLs using a more structured approach, since the builders will guarantee that a slash will be added when joining a domain with a path and they usually also prevent double slashes (e.g. `https://example.com//.well-known/jwks.json), which can sometimes mess with string comparisons of URLs that are equivalent but not string-equal.

If it sounds like a reasonable change, do you think you change this so that we can use the URI class more? The docs aren't great, but I think either you can have this method return a `URI` object, and then have the callers append a new path to it if they need to, or you can just have this method only return the domain, without the `https://` or the trailing `/`, and then have the callers construct the `URI` objects by separately passing in the `host:` and `path:` keys.

Reply:
Sure, that trailing slash certainly makes things brittle! I changed this to construct the base URL using the URI:HTTPS library, which I assign to a class constant rather than constructing it within a method. I then append paths where needed using the URI.join method